src/index/actor.rs

```diff
@@ -7,10 +7,40 @@ use crate::Distance;
 use crate::Embedding;
 use crate::Limit;
 use crate::PrimaryKey;
+use std::fmt;
 use tokio::sync::mpsc;
 use tokio::sync::oneshot;
 
-pub(crate) type AnnR = anyhow::Result<(Vec<PrimaryKey>, Vec<Distance>)>;
+#[derive(Debug)]
+pub enum AnnError {
+    WrongEmbeddingDimension { expected: usize, actual: usize },
+    OtherError(anyhow::Error),
+}
+
+impl<E> From<E> for AnnError
+where
+    E: std::error::Error + Send + Sync + 'static,
+{
+    fn from(err: E) -> Self {
+        AnnError::OtherError(anyhow::anyhow!(err))
+    }
+}
+
+impl fmt::Display for AnnError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            AnnError::WrongEmbeddingDimension { expected, actual } => {
+                write!(
+                    f,
+                    "Wrong embedding dimension: expected {expected}, got {actual}"
+                )
+            }
+            AnnError::OtherError(err) => write!(f, "Other error: {err}"),
+        }
+    }
+}
+
+pub(crate) type AnnR = anyhow::Result<(Vec<PrimaryKey>, Vec<Distance>), AnnError>;
```

Comment:
```suggestion
pub(crate) type AnnR = anyhow::Result<(Vec<PrimaryKey>, Vec<Distance>)>;
```

I would propose to use [anyhow::bail!](https://docs.rs/anyhow/latest/anyhow/macro.bail.html) to send custom error as anyhow::Error

Reply:
Returning custom errors is mostly done on libraries (crates). On the application side using anyhow::Error is a common path. We could think how to refactor error maintenance as we need a way to log errors, send back http errors. But for this we need to get some experience what is really needed in this application.